### PR TITLE
Fix `IndexDefect` in `httpclient` and fix parsing of multiline headers

### LINKED ISSUE #19261: `IndexDefect` in httpclient because of a header spanning multiple lines
I was doing some scanning stuff and found a host that replies with one header that spans multiple lines. Chromium seems to be happy with it, but `httpclient` does an out of bounds access which results in IndexDefect.

### Example
No usable-by-default Nim repro code yet, but here's the whole response from the server:
```html
> GET / HTTP/1.1
> Host: [REDACTED]
> User-Agent: curl/7.80.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx
< Date: Fri, 17 Dec 2021 00:23:58 GMT
< Content-Type: text/html
< Content-Length: 178
< Connection: keep-alive
< Keep-Alive: timeout=5
< Location: ./login.html
< Content-Security-Policy: default-src 'self';
<                                     script-src 'self'
<                                                'unsafe-inline'
<                                                'unsafe-eval';
<                                     img-src 'self'
<                                             data:;
<                                     style-src 'self'
<                                               'unsafe-inline';
<                                     font-src 'self';
<                                     child-src 'self';
<                                     object-src 'none'
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-XSS-Protection: 1; mode=block
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< 
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

Notice how `Content-Security-Policy` spans multiple lines.

### Current Output
```
/home/dian/Projects/PrivWeb/scratchpad.nim(47) scratchpad
/home/dian/Things/nim/lib/pure/httpclient.nim(1188) getContent
/home/dian/Things/nim/lib/pure/httpclient.nim(1183) get
/home/dian/Things/nim/lib/pure/httpclient.nim(1110) request
/home/dian/Things/nim/lib/pure/httpclient.nim(1057) requestAux
/home/dian/Things/nim/lib/pure/httpclient.nim(846) parseResponse
/home/dian/Things/nim/lib/system/fatal.nim(53) sysFatal
Error: unhandled exception: index 53 not in 0 .. 52 [IndexDefect]
```

### Expected Output
No error from `httpclient`, and it should parse that header into a single line, like Chromium:
image

### Additional Information
```
$ nim -v
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2021-12-17
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 0d0c249074d6a1041de16108dc247396efef5513
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ghost**
Follow the part of the HTTP spec regarding headers - this solves the actual issue in #19261 as well as makes `httpclient` follow the spec more closely. I've chosen not to enforce the spec to the level Firefox does (regarding disallowing spaces after the header name but before the value) as I've wanted to limit this change just to fix the bug that I encountered.

Also, I don't know how I should make the tests for this.

### CODE REVIEW

File: lib/pure/httpclient.nim [hunk 1]
```diff
@@ -840,13 +841,23 @@ proc parseResponse(client: HttpClient | AsyncHttpClient,
...
+        inc(linei, le)
+        if line[linei] != ':': httpError("invalid headers")
+        inc(linei) # Skip :
+
+        result.headers.add(name, line[linei .. ^1].strip())
+      # If there are spaces before the header name, it's actually a value
+      # that should be appended to the previous header, see bug #19261
+      # Also, if there was no header before this, we just ignore the line
+      elif prevHeader != "":
+        result.headers.table[prevHeader][^1].add line.strip()
```
Comment (Yardanico) [inline lib/pure/httpclient.nim:859]: We know that `prevHeader` is already in the table because we check if it's empty or not.

**Comment by ghost** [comment]
Also, we might want to check that `linei` is actually less than `line.len` just to be safe, should I do it in this PR?

**Comment by ghost** [comment]
~~Okay, I forgot that httpclient saves headers in a different case than the default one, so my branch will error out in most cases. I'm thinking of a fix, but the problem is that `toCaseInsensitive` from `httpcore` is private, so I can either export it or copy it to httpclient.~~

Currently fixed by exporting `toCaseInsensitive` just to make a PoC, I don't know if there's a better way than this.

File: lib/pure/httpclient.nim [hunk 2]
```diff
@@ -840,13 +841,21 @@ proc parseResponse(client: HttpClient | AsyncHttpClient,
...
       var name = ""
-      var le = parseUntil(line, name, ':', linei)
-      if le <= 0: httpError("invalid headers")
-      inc(linei, le)
-      if line[linei] != ':': httpError("invalid headers")
-      inc(linei) # Skip :
+      var leadingSpaces = skipWhitespace(line, linei)
+      if leadingSpaces == 0:
+        var le = parseUntil(line, name, ':', linei)
+        if le <= 0 or le >= line.len: httpError("invalid headers")
```
Comment (ghost) [inline lib/pure/httpclient.nim]: `le >= line.len` makes `httpclient` a little more stricter regarding random lines it finds in the response - basically, if the response is like:
```
Header-One: value
Header-Two: value
somerandomname
```

`httpclient` will error out on `somerandomname`. Chromium and Firefox don't error out on that and just skip the line - should we follow them or throw an error?

File: lib/pure/httpclient.nim [hunk 3]
```diff
@@ -840,13 +841,21 @@ proc parseResponse(client: HttpClient | AsyncHttpClient,
...
+        prevHeader = name
+        inc(linei, le)
+        if line[linei] != ':': httpError("invalid headers")
+        inc(linei) # Skip :
+        result.headers.add(name, line[linei .. ^1].strip())
+      # If there are spaces before the header name, it's actually a value
+      # that should be appended to the previous header, see bug #19261
+      # Also, if there was no header before this, we just ignore the line
+      elif prevHeader != "":
+        result.headers.table[result.headers.toCaseInsensitive(prevHeader)][^1].add line.strip()
```
Comment (ghost) [inline lib/pure/httpclient.nim]: We know that prevHeader is already in the table because we check if it's empty or not.

**Comment by ghost** [comment]
I'll see if I can either try to test this PR more properly to make it ready to merge, or make a smaller PR just to fix the IndexDefect without changing anything else. I hope we can get this into 2.0, sorry for not doing it before

**Comment by ghost** [comment]
Going to open a new PR instead

**Pull request closed without merging**